### PR TITLE
Space - Becca

### PR DESCRIPTION
**Pull request by RebeccaRoach**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. |  As a helper method, I wrote a ```most_recent_vote_date``` custom model instance method on the works model to use in the spotlight method in case of a tie for top voted work. The ```most_recent_vote_date``` method returns the most recent date on which a user voted, which made it possible to implement tie-breaking logic per the demo site.
Describe how you approached testing that model method. What edge cases did you come up with? | In testing this method, I explicitly added different created_at attributes in my ```votes.yml``` file for two different votes on the same work. This allowed me to see that the most recent vote date was returned instead of the earlier one. As an edge case, I tested what would happen if a work did not have any votes, in which case I return an old dummy date (May 20, 1965). Doing this allowed me to still do the comparison of dates even if a work didn't have a vote, vs. trying to compare a date with nil.
What are `session` and `flash`? What is the difference between them? | Session and flash are hash-like objects that allow us to persist data in rails between requests. The main difference is that flash data will persist only throughout the next request-response cycle (or sooner with flash.now), while session persists data until the session is cleared, whether by logging out, closing the browser, clearing cookies, or another similar user action.
What was one thing that you gained more clarity on through this assignment? | I gained clarity on the possibilities yet dangers of attempting complicated database queries inside views. In trying to call my custom model methods within the show and index views at first, I learned that it's expensive and sometimes confusing to try to access data like this (and that it could lead to poor performance in the long run, especially with lots of data on different media to account for, etc.) Instead, I learned the importance of exposing instance variables from the controller actions that you can then access in the views. 
What is the Heroku URL of your deployed application? | https://beccas-media-ranker-clone.herokuapp.com/

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  get '/login', to: "users#login_form", as: "login"
+  post '/login', to: "users#login"
+  post '/logout', to: "users#logout", as: "logout"
+
+  resources :works
+  post '/works/:id/upvote', to: 'works#upvote', as: "upvote"
+
+  resources :users, only: [:index, :show, :new]
+
+  get '/homepages', to: 'homepages#index', as: 'homepages'
```
Comment (CheezItMan) [inline config/routes.rb:12]: Since you have the root path going to homepages#index, why do you need this additional route?

File: test/models/vote_test.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    it "is not a valid vote if work/work_id is missing" do
+      attempted_vote = Vote.new(user_id: 1)
+      result = attempted_vote.save
+  
+      expect(result).must_equal false
+      expect(Vote.all.count).must_equal @previous_vote_count
+    end
+
+    it "is invalid for a user to vote more than once on the same work" do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:86]: 👍

File: test/controllers/homepages_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe HomepagesController do
+  # it "does a thing" do
```
Comment (CheezItMan) [inline test/controllers/homepages_controller_test.rb:4]: Just noting no tests here.

Comment (CheezItMan) [reply test/controllers/homepages_controller_test.rb:4]: Also no works_controller_test

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class ApplicationController < ActionController::Base
+  rescue_from ActiveRecord::RecordNotFound, with: :record_not_found
+
+  def record_not_found
+    render file: "#{Rails.root}/public/404.html" , status: 404
+  end
```
Comment (CheezItMan) [inline app/controllers/application_controller.rb:6]: Nice!

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  def create
+    @work = Work.new(work_params)
+
+    if @work.save
+      flash[:success] = "Successfully created #{@work.category} #{@work.id}"
+      redirect_to @work
+    else
+      flash[:error] = "A problem occurred: Could not create #{@work.category}"
+      flash[:error_messages] = @work.errors.messages
+      render :new
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:33]: You should also include a status code

```suggestion
      render :new, status: :bad_request
```

File: app/controllers/works_controller.rb [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  end
+
+  def update
+    if @work.update(work_params)
+      flash[:success] = "Successfully updated #{@work.category} #{@work.id}"
+      redirect_to @work
+    else
+      flash[:error] = "A problem occurred: Could not update #{@work.category}"
+      flash[:error_messages] = @work.errors.messages
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:44]:
```suggestion
      render :edit, status: :bad_request
```

File: app/models/work.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+  def self.get_in_category_by_votes(category_name)
+    valid_categories = ["album", "book", "movie"]
+    raise ArgumentError.new("#{category_name} is not valid") if !valid_categories.include?(category_name)
+
+    num_category_works = self.where(category: category_name).count
+
+    return self.where(category: category_name).includes(:votes).max_by(num_category_works) {
+      |work| work.votes.count
+    }
```
Comment (CheezItMan) [inline app/models/work.rb:17]: Just noting it's better to sort the works at the database-level with things like `.order` as Postgres is more efficient at filtering and sorting records than Ruby/Rails are.

**Comment by CheezItMan** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️, you're using the default 404 page.
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️?
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️?
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️?

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️, really good use of partials here
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅


## Summary


Nicely done, this is really outstanding work.  You hit all the learning goals.  Well done.  Take a look at my comments and let me know what questions you have.
